source/common/network/buffered_io_socket_handle_impl.cc

```diff
@@ -0,0 +1,228 @@
+#include "common/network/buffered_io_socket_handle_impl.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/platform.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/common/assert.h"
+#include "common/common/utility.h"
+#include "common/event/file_event_impl.h"
+#include "common/network/address_impl.h"
+
+#include "absl/container/fixed_array.h"
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Network {
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::close() {
+  ASSERT(!closed_);
+  if (!write_shutdown_) {
+    ASSERT(writable_peer_);
+    // Notify the peer we won't write more data. shutdown(WRITE).
+    writable_peer_->setWriteEnd();
+    // Notify the peer that we no longer accept data. shutdown(RD).
+    writable_peer_->onPeerDestroy();
+    writable_peer_->maybeSetNewData();
+    writable_peer_ = nullptr;
+    write_shutdown_ = true;
+  }
+  closed_ = true;
+  return Api::ioCallUint64ResultNoError();
+}
+
+bool BufferedIoSocketHandleImpl::isOpen() const { return !closed_; }
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::readv(uint64_t max_length,
+                                                          Buffer::RawSlice* slices,
+                                                          uint64_t num_slice) {
+  if (owned_buffer_.length() == 0) {
+    if (read_end_stream_) {
+      return {0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+    } else {
+      return {0, Api::IoErrorPtr(IoSocketError::getIoSocketEagainInstance(),
+                                 IoSocketError::deleteIoError)};
+    }
+  } else {
+    absl::FixedArray<iovec> iov(num_slice);
+    uint64_t num_slices_to_read = 0;
+    uint64_t num_bytes_to_read = 0;
+    for (; num_slices_to_read < num_slice && num_bytes_to_read < max_length; num_slices_to_read++) {
+      auto min_len = std::min(std::min(owned_buffer_.length(), max_length) - num_bytes_to_read,
+                              uint64_t(slices[num_slices_to_read].len_));
+      owned_buffer_.copyOut(num_bytes_to_read, min_len, slices[num_slices_to_read].mem_);
```

Comment:
you could also drain as you go and not have to worry about offset and slice index here

Reply:
`readv(slices)` doesn't drain.
`read(buffer)` does and is implemented in this PR.
I think readv will be deprecated soon now that we have read(buffer)